File: sdk/packages/python/iii-example/src/hooks.py
from typing import Any, Awaitable, Callable

from iii import ApiRequest, ApiResponse, IIIClient, Logger
from iii.iii_types import FunctionInfo


def use_api(
    iii: IIIClient,
    config: dict[str, Any],
    handler: Callable[[ApiRequest[Any], Logger], Awaitable[ApiResponse[Any]]],
) -> None:
    api_path = config["api_path"]
    http_method = config["http_method"]
    function_id = f"api.{http_method.lower()}.{api_path}"
    logger = Logger(service_name=function_id)

    async def wrapped(data: ApiRequest) -> dict[str, Any]:
        req = ApiRequest(**data) if isinstance(data, dict) else data
        result = await handler(req, logger)
        return result.model_dump(by_alias=True)

    iii.register_function(function_id, wrapped)
    iii.register_trigger(
        {
            "type": "http",
            "function_id": function_id,
            "config": {
                "api_path": api_path,
                "http_method": http_method,
                "description": config.get("description"),
                "metadata": config.get("metadata"),
            },
        }
    )


def use_functions_available(iii: IIIClient, callback: Callable[[list[FunctionInfo]], None]) -> Callable[[], None]:
    return iii.on_functions_available(callback)

File: sdk/packages/node/iii-example/src/hooks.ts
import { type ApiRequest, type ApiResponse, Logger } from 'iii-sdk'
import { iii } from './iii'

// biome-ignore lint/suspicious/noExplicitAny: generic default requires any for handler flexibility
export const useApi = <TBody = any>(
  config: {
    api_path: string
    http_method: string
    description?: string
    metadata?: Record<string, unknown>
  },
  handler: (req: ApiRequest<TBody>, logger: Logger) => Promise<ApiResponse>,
) => {
  const function_id = `api::${config.http_method.toLowerCase()}::${config.api_path}`
  const logger = new Logger(undefined, function_id)

  iii.registerFunction(function_id, req => handler(req, logger), { metadata: config.metadata })
  iii.registerTrigger({
    type: 'http',
    function_id,
    config: {
      api_path: config.api_path,
      http_method: config.http_method,
      description: config.description,
      metadata: config.metadata,
    },
  })
}
